File: stack-n-queue/sol/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> d;
        vector<int> ans;
        for(int i = 0;i < nums.size(); i++){
            // pushing into deque.
            while(!d.empty() && nums[i] >= nums[d.back()]){
                d.pop_back();
            }
            d.push_back(i);


            // checking if the window is valid or not...if not then remove one extra element.
            if(d.back() - d.front() + 1 > k){
                d.pop_front();
            }


            // When the first window is made .. now we can ans the query.
            if(i >=  k-1){
                ans.push_back(nums[d.front()]);
            }
        }
        return ans;
    }
};

File: stack-n-queue/sol/sliding-window-maximum.java
import java.util.*;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> d = new ArrayDeque<>();
        int n = nums.length;
        int[] ans = new int[n - k + 1];
        int idx = 0;

        for (int i = 0; i < n; i++) {

            // pushing into deque (remove smaller elements from back)
            while (!d.isEmpty() && nums[i] >= nums[d.peekLast()]) {
                d.pollLast();
            }
            d.offerLast(i);

            // remove elements outside window
            if (d.peekFirst() <= i - k) {
                d.pollFirst();
            }

            // first window formed -> record answer
            if (i >= k - 1) {
                ans[idx++] = nums[d.peekFirst()];
            }
        }

        return ans;
    }
}
